src/BatteryStateOfChargeService.cpp

```diff
@@ -26,38 +26,70 @@
  */
 
 #include "BatteryStateOfChargeService.h"
+#include "BatteryData.h"
+
 
 namespace
 {
     const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+    const double MSECS_TO_HOURS=2.77778e-7;
 }
 
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
-: initialStateOfChargePercent_(initialStateOfChargePercent)
+: initialStateOfChargePercent_(initialStateOfChargePercent),newCurrent_(0), sumCurrent_(0)
 {
+ AmpHoursUsed_=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/100);
 }
-
 BatteryStateOfChargeService::~BatteryStateOfChargeService()
 {
 }
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return (AmpHoursUsed_);
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(newCurrent_>=0)
     return false;
+    else
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+  double timeLeft;
+  timeLeft=(AmpHoursUsed_/newCurrent_);
+  timeLeft=qAbs(timeLeft);
+  timeLeft=timeLeft/MSECS_TO_HOURS;
+
+  QTime base(0,0);
+  return base.addMSecs(timeLeft);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+ //Setting Values
+  double inCurrent_=newCurrent_;
+   QTime previous_Time=currentTime_;
+
+    newCurrent_=batteryData.current;
+    currentTime_=batteryData.time;
+
+    if(firstRun_!=true){
```

Comment:
So it's generally better to write an `if` statement for special cases, instead of writing it for the default. So you should move all the logic inside this `if` statement outside, and move the special case logic into the `if` statement.